org.eclipse.xtext.ide/src/org/eclipse/xtext/ide/server/WorkspaceManager.java

```diff
@@ -141,16 +149,68 @@ public void setBuildManager(BuildManager buildManager) {
 	 */
 	public void initialize(URI baseDir, Procedure2<? super URI, ? super Iterable<Issue>> issueAcceptor,
 			CancelIndicator cancelIndicator) {
-		this.baseDir = baseDir;
+		if (isSupportsWorkspaceFolders()) {
+			WorkspaceFolder workspaceFolder = new WorkspaceFolder(uriExtensions.toUriString(baseDir), "workspace");
+			initialize(Collections.singletonList(workspaceFolder), issueAcceptor, cancelIndicator);			
+		} else {
+			this.baseDir = baseDir;
+			this.issueAcceptor = issueAcceptor;
+			refreshWorkspaceConfig(cancelIndicator);
+		}
+	}
+
+	/**
+	 * Initialize a workspace with the given workspace folders.
+	 *
+	 * @param workspaceFolders
+	 *            the list of workspace root folders
+	 * @param issueAcceptor
+	 *            the issue acceptor
+	 * @param cancelIndicator
+	 *            allows to cancel the initialization
+	 * @since 2.21
+	 */
+	public void initialize(List<WorkspaceFolder> workspaceFolders, Procedure2<? super URI, ? super Iterable<Issue>> issueAcceptor,
+			CancelIndicator cancelIndicator) {
+		this.workspaceFolders = new ArrayList<>(workspaceFolders);
 		this.issueAcceptor = issueAcceptor;
 		refreshWorkspaceConfig(cancelIndicator);
 	}
 
+	/**
+	 * @return whether this workspace manager supports multiple workspace root folders.
+	 * @since 2.21
+	 */
+	public boolean isSupportsWorkspaceFolders() {
+		return workspaceConfigFactory instanceof IMultiRootWorkspaceConfigFactory;
```

Comment:
See question above on the interface `IMultiRootWorkspaceConfigFactory` - this code here suggests that the "normal" workspaceConfigFactory and the multi root incarnation of it are in an inheritance relationship.

Reply:
Changed it.